frame/node-permission/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+
+        type Error = Error<T>;
+
+        fn deposit_event() = default;
+
+        /// Add a node to the allowlist.
+        ///
+        /// May only be called from `T::AddOrigin`.
+        ///
+        /// - `node_id`: identifier of the node, it's likely the public key of ed25519 keypair.
+        #[weight = (T::WeightInfo::add_node(), DispatchClass::Operational)]
+        pub fn add_node(origin, node_id: T::NodeId) {
+            T::AddOrigin::ensure_origin(origin)?;
+
+            let mut nodes = Self::get_allowlist();
+            ensure!(nodes.len() < T::MaxPermissionedNodes::get() as usize, Error::<T>::TooManyNodes);
+
+            let location = nodes.binary_search(&node_id).err().ok_or(Error::<T>::AlreadyJoined)?;
+            nodes.insert(location, node_id);
+            Self::put_allowlist(nodes);
+
+            Self::deposit_event(Event::NodeAdded);
+        }
+
+        /// Remove a node from the allowlist.
+        ///
+        /// May only be called from `T::RemoveOrigin`.
+        ///
+        /// - `node_id`: identifier of the node, it's likely the public key of ed25519 keypair.
+        #[weight = (T::WeightInfo::remove_node(), DispatchClass::Operational)]
+        pub fn remove_node(origin, node_id: T::NodeId) {
+            T::RemoveOrigin::ensure_origin(origin)?;
+
+            let mut nodes = Self::get_allowlist();
+
+            let location = nodes.binary_search(&node_id).ok().ok_or(Error::<T>::NotExist)?;
+            nodes.remove(location);
+            Self::put_allowlist(nodes);
+
+            Self::deposit_event(Event::NodeRemoved);
+        }
+
+        /// Swap two nodes.
+        ///
+        /// May only be called from `T::SwapOrigin`.
+        ///
+        /// - `remove`: the node which will be moved out from the list, it's likely the public key
+        /// of ed25519 keypair.
+        /// - `add`: the node which will be put in the list, it's likely the public key of ed25519
+        /// keypair.
+        #[weight = (T::WeightInfo::swap_node(), DispatchClass::Operational)]
+        pub fn swap_node(origin, remove: T::NodeId, add: T::NodeId) {
+            T::SwapOrigin::ensure_origin(origin)?;
+
+            if remove == add { return Ok(()) }
+            
+            let mut nodes = Self::get_allowlist();
+            let location = nodes.binary_search(&remove).ok().ok_or(Error::<T>::NotExist)?;
+            let _ = nodes.binary_search(&add).err().ok_or(Error::<T>::AlreadyJoined)?;
+            nodes[location] = add;
+            nodes.sort();
+            Self::put_allowlist(nodes);
+
+            Self::deposit_event(Event::NodesSwapped);
+        }
+
+        /// Reset all the permissioned nodes in the list.
+        ///
+        /// May only be called from `T::ResetOrigin`.
+        ///
+        /// - `nodes`: the new nodes for the allowlist.
+        #[weight = (T::WeightInfo::reset_nodes(), DispatchClass::Operational)]
+        pub fn reset_nodes(origin, nodes: Vec<T::NodeId>) {
```

Comment:
```suggestion
        pub fn reset_nodes(origin, mut nodes: Vec<T::NodeId>) {
```

Reply:
Here `mut` breaks the compiler, I guess we are not allowed to use in the macro. @bkchr